src/psyclone/tests/domain/lfric/dofkern_test.py

```diff
@@ -142,3 +152,98 @@ def test_dof_kernel_invalid_field_vector():
     assert ("Kernel 'testkern_dofs_type' operates on 'dof' but has a vector "
             "argument 'gh_field*3'. This is not permitted in the LFRic API."
             in str(excinfo.value))
+
+
+def test_upper_bounds(monkeypatch, annexed, dist_mem, tmpdir):
+    '''
+    Checks that the correct upper bound is generated for a dof-kernel for all
+    permutations of the `DISTRIBUTED_MEMORY` and `COMPUTE_ANNEXED_DOFS`
+    configuration settings.
+
+    '''
+    # Set up annexed dofs
+    config = Config.get()
+    lfric_config = config.api_conf("lfric")
+    monkeypatch.setattr(lfric_config, "_compute_annexed_dofs", annexed)
+
+    _, invoke_info = parse(os.path.join(BASE_PATH,
+                                        "1.14_single_invoke_dofs.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=dist_mem).create(invoke_info)
+    code = str(psy.gen)
+
+    # Distributed memory
+    if annexed and dist_mem:
+        expected = ("      loop0_start = 1\n"
+                    "      loop0_stop = f1_proxy%vspace%get_last_dof_annexed()"
+                    )
+    elif not annexed and dist_mem:
+        expected = ("      loop0_start = 1\n"
+                    "      loop0_stop = f1_proxy%vspace%get_last_dof_owned()"
+                    )
+
+    # Shared memory
```

Comment:
Annexed only has any effect if distributed-memory is enabled so you can simplify this to `elif not dist_mem:`